File: shared_rs/src/domain/pipeline_runner.rs
//! Shared domain: pipeline runner.
//!
//! A [`Pipeline`] is a named, ordered sequence of [`Stage`] instances.

use std::time::Instant;

use super::pipeline::{PipelineContext, Stage, StageResult};

/// Named, ordered sequence of stages.
pub struct Pipeline {
    pub name: String,
    stages: Vec<Box<dyn Stage>>,
}

impl Pipeline {
    pub fn new(name: impl Into<String>, stages: Vec<Box<dyn Stage>>) -> Self {
        Self {
            name: name.into(),
            stages,
        }
    }

    /// Execute every stage in order, recording timing in `ctx.stage_results`.
    pub fn run(&self, mut ctx: PipelineContext) -> anyhow::Result<PipelineContext> {
        for stage in &self.stages {
            let t0 = Instant::now();
            ctx = stage.process(ctx)?;
            ctx.stage_results.push(StageResult {
                stage: stage.name().to_owned(),
                elapsed_secs: t0.elapsed().as_secs_f64(),
            });
        }
        Ok(ctx)
    }

    /// Load heavy resources for every stage.
    pub fn load_all(&mut self) -> anyhow::Result<()> {
        for stage in &mut self.stages {
            stage.load()?;
        }
        Ok(())
    }

    /// Free resources for every stage.
    pub fn unload_all(&mut self) -> anyhow::Result<()> {
        for stage in &mut self.stages {
            stage.unload()?;
        }
        Ok(())
    }
}

File: tts_python/domain/pipeline_runner.py
"""
Shared domain: pipeline runner.

A Pipeline is a named, ordered sequence of Stage instances.
"""

from __future__ import annotations

import time

from shared.domain.pipeline import PipelineContext, Stage


class Pipeline:
    """Named, ordered sequence of stages."""

    def __init__(self, name: str, stages: list[Stage]) -> None:
        self.name = name
        self.stages = stages

    def run(self, ctx: PipelineContext) -> PipelineContext:
        """Execute every stage in order, recording timing in *ctx.stage_results*."""
        for stage in self.stages:
            t0 = time.time()
            ctx = stage.process(ctx)
            ctx.stage_results.append(
                {
                    "stage": stage.name,
                    "elapsed": time.time() - t0,
                }
            )
        return ctx

    def load_all(self) -> None:
        """Load heavy resources for every stage."""
        for s in self.stages:
            s.load()

    def unload_all(self) -> None:
        """Free resources for every stage."""
        for s in self.stages:
            s.unload()


__all__ = ["Pipeline"]
